### PR TITLE
ZCS-10060 :  Replaced Ganymed SSH-2 with Apache MINA SSHD

### PR DESCRIPTION
**Pull request by umagmrit**
Problem: RemoteManager calls are failing with FIPS enabled.
https://jira.corp.synacor.com/browse/ZCS-10060?focusedCommentId=1419982&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-1419982

Ganymed SSH-2 offering KexAlgorithms: diffie-hellman-group-exchange-sha1, diffie-hellman-group14-sha1, diffie-hellman-group1-sha1 , which are not allowed in FIPS mode.

Solution : Replaced Ganymed SSH-2 with Apache MINA SSHD , which supports KexAlgorithms allowed in FIPS mode.

### CODE REVIEW

**Review by chinmay15 — APPROVED** [review]
LGTM

File: store/src/java/com/zimbra/cs/rmgmt/RemoteManager.java [hunk 1]
```diff
@@ -161,54 +154,72 @@ public synchronized RemoteResult execute(String command) throws ServiceException
...
+				.verify(defaultTimeoutSeconds, TimeUnit.SECONDS).getSession()) {
+			session.addPublicKeyIdentity(loadKeypair(privateKey.getAbsolutePath()));
+			session.auth().verify(defaultTimeoutSeconds, TimeUnit.SECONDS);
+			if (ZimbraLog.rmgmt.isDebugEnabled()) {
+				ZimbraLog.rmgmt.debug("executing shim command '%s'", mShimCommand);
+			}
+			try (ByteArrayOutputStream responseStream = new ByteArrayOutputStream();
+					ByteArrayOutputStream errorResponseStream = new ByteArrayOutputStream();
+					ClientChannel channel = session.createChannel("exec", mShimCommand + "\n")) {
+				if (ZimbraLog.rmgmt.isDebugEnabled()) {
```
Comment (Prashantsurana) [inline store/src/java/com/zimbra/cs/rmgmt/RemoteManager.java:180]: no need to if block logger should handle it.

File: store/src/java/com/zimbra/cs/rmgmt/RemoteManager.java [hunk 2]
```diff
@@ -161,54 +154,72 @@ public synchronized RemoteResult execute(String command) throws ServiceException
...
+					result.mStderr = ByteUtil.getContent(stderr, -1);
+					result.mExitStatus = channel.getExitStatus();
+					if (result.mExitStatus != 0) {
+						throw new IOException("command failed: exit status=" + result.mExitStatus + ", stdout="
+								+ new String(result.mStdout) + ", stderr=" + new String(result.mStderr));
+					}
+					result.mExitSignal = channel.getExitSignal();
+					String errorString = new String(errorResponseStream.toByteArray());
+					if (!errorString.isEmpty()) {
+						throw new Exception(errorString);
```
Comment (Prashantsurana) [inline store/src/java/com/zimbra/cs/rmgmt/RemoteManager.java:203]: throw ServiceException

File: store/src/java/com/zimbra/cs/rmgmt/RemoteManager.java [hunk 3]
```diff
@@ -95,27 +106,25 @@ public Integer getPort() {
...
-            String send = "HOST:" + mHost + " " + command;
-            if (ZimbraLog.rmgmt.isDebugEnabled()) ZimbraLog.rmgmt.debug("(bg) sending mgmt command '" + send + "' on " + this);
-            os.write(send.getBytes());
-            os.close();
-            InputStream stdout = new StreamGobbler(s.getStdout());
-            InputStream stderr = new StreamGobbler(s.getStderr());
-            handler.read(stdout, stderr);
-        } catch (OutOfMemoryError e) {
+            result = executeRemoteCommand(mUser,mHost,mPort,mPrivateKey,mShimCommand,command);
+            if (ZimbraLog.rmgmt.isTraceEnabled()) {
```
Comment (Prashantsurana) [inline store/src/java/com/zimbra/cs/rmgmt/RemoteManager.java:112]: do we need check here?

File: store/src/java/com/zimbra/cs/rmgmt/RemoteManager.java [hunk 4]
```diff
@@ -161,54 +154,72 @@ public synchronized RemoteResult execute(String command) throws ServiceException
...
         }
+        return result;
     }
 
-    private Connection mConnection;
-
-    private void releaseSession(Session sess) {
-        try {
-            sess.close();
-        } finally {
```
Comment (Prashantsurana) [inline store/src/java/com/zimbra/cs/rmgmt/RemoteManager.java:191]: are we closing the connection anywhere after error?

**Review by Prashantsurana — CHANGES_REQUESTED** [review]

**Pull request merged**